### PR TITLE
#320 Make the data "per user"

### LINKED ISSUE #320: Make the data "per user"
**Is your feature request related to a problem? If so, please describe.**
Some time ago, users/login/registration was added to the application.
Yet, the book/shelf logic and relations were not updated for multi-user purposes.
Every user currently shares the same data.

**Describe the solution you'd like**
Add some kind of ownership relation to users/books/shelves.
Each user should have their own books and shelves.

**Additional context**
I'm not sure how this works out with the predefined shelves, maybe the shelves should be generalized into a "shelf" entity and the predefined ones are automatically created upon registration, or some other solution.

### PR DESCRIPTION
**Pull request by Burhak**
## Summary of change

- added relation to `User` to `Shelf` entities
- select queries on shelves now have condition on user
- creating new instances of shelves is in competence of according shelf service (I suggest to keep it this way in future)
- insertion of test data is now made per user
- static initialization of tests is refactored due to `@WithMockedUser`taking effect after initialization of application context

## Related issue

Closes #320 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

### CODE REVIEW

**Comment by knjk04** [comment]
@Burhak Thanks for this! We'll review it shortly

File: src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java [hunk 1]
```diff
@@ -19,10 +19,12 @@
...
 
 import com.karankumar.bookproject.backend.entity.CustomShelf;
+import com.karankumar.bookproject.backend.entity.account.User;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 import java.util.List;
 
 public interface CustomShelfRepository extends JpaRepository<CustomShelf, Long> {
-    List<CustomShelf> findByShelfName(String shelfName);
+    List<CustomShelf> findAllByUser(User user);
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java:28]: Do we really need this method? What would be the use case for this method?

File: src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java [hunk 2]
```diff
@@ -19,10 +19,12 @@
...
+    CustomShelf findByShelfNameAndUser(String shelfName, User user);
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java:29]: Could you please add test for this?

File: src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java [hunk 3]
```diff
@@ -19,8 +19,13 @@
...
 
 import com.karankumar.bookproject.backend.entity.PredefinedShelf;
+import com.karankumar.bookproject.backend.entity.account.User;
 import org.springframework.data.jpa.repository.JpaRepository;
 
+import java.util.List;
+
 public interface PredefinedShelfRepository extends JpaRepository<PredefinedShelf, Long> {
-    PredefinedShelf findByPredefinedShelfName(PredefinedShelf.ShelfName shelfName);
+    List<PredefinedShelf> findAllByUser(User user);
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java:28]: Do we really need this method? What would be the use case for this method?

File: src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java [hunk 4]
```diff
@@ -19,8 +19,13 @@
...
+    PredefinedShelf findByPredefinedShelfNameAndUser(PredefinedShelf.ShelfName shelfName, User user);
+    int countAllByUser(User user);
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java:30]: Could you please add test for this?

File: src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java [hunk 5]
```diff
@@ -39,15 +45,11 @@ public CustomShelf findById(Long id) {
     }
 
     public List<CustomShelf> findAll() {
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java:47]: Renaming the method?

File: src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java [hunk 6]
```diff
@@ -39,15 +45,11 @@ public CustomShelf findById(Long id) {
...
-        return customShelfRepository.findAll();
+        return customShelfRepository.findAllByUser(userService.getCurrentUser());
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java:48]: Could you please add test for this?

File: src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java [hunk 7]
```diff
@@ -39,15 +45,11 @@ public CustomShelf findById(Long id) {
...
     }
 
-    public List<CustomShelf> findAll(String shelfName) {
-        if (shelfName == null) {
-            return customShelfRepository.findAll();
-        } else {
-            return customShelfRepository.findByShelfName(shelfName);
-        }
+    public CustomShelf findByShelfName(String shelfName) {
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java:51]: Could you please add test for this?

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java [hunk 8]
```diff
@@ -44,14 +46,16 @@
     private final PredefinedShelfRepository predefinedShelfRepository;
     private final AuthorRepository authorRepository;
     private final TagRepository tagRepository;
+    private final UserService userService;
 
     public PredefinedShelfService(BookRepository bookRepository, AuthorRepository authorRepository,
-            PredefinedShelfRepository shelfRepository, TagRepository tagRepository) {
+            PredefinedShelfRepository shelfRepository, TagRepository tagRepository, UserService userService) {
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:52]: Do we really need `UserService` dependency in `ShelfService`? All we need from `UserService` is the current user so wouldn't it be better to have this at the repo layer? There may not be any change at the service layer and we pass the current logged in user to shelf repo?

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java [hunk 9]
```diff
@@ -70,27 +74,27 @@ public void save(PredefinedShelf shelf) {
     }
 
     public List<PredefinedShelf> findAll() {
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:76]: We can rename this method `findAllForLoggedInUser`?

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java [hunk 10]
```diff
@@ -70,27 +74,27 @@ public void save(PredefinedShelf shelf) {
...
-        return predefinedShelfRepository.findAll();
+        return predefinedShelfRepository.findAllByUser(userService.getCurrentUser());
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:77]: Could you please add test for this?

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java [hunk 11]
```diff
@@ -70,27 +74,27 @@ public void save(PredefinedShelf shelf) {
...
-        return predefinedShelfRepository.findByPredefinedShelfName(PredefinedShelf.ShelfName.READ);
+        return findByPredefinedShelfName(PredefinedShelf.ShelfName.READ);
     }
 
     public PredefinedShelf findDidNotFinishShelf() {
-        return predefinedShelfRepository.findByPredefinedShelfName(PredefinedShelf.ShelfName.DID_NOT_FINISH);
+        return findByPredefinedShelfName(PredefinedShelf.ShelfName.DID_NOT_FINISH);
     }
 
     public PredefinedShelf findByPredefinedShelfName(PredefinedShelf.ShelfName shelfName) {
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:96]: We can rename this method `findByPredefinedShelfNameAndLoggedInUser`?

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java [hunk 12]
```diff
@@ -70,27 +74,27 @@ public void save(PredefinedShelf shelf) {
...
-        return predefinedShelfRepository.findByPredefinedShelfName(shelfName);
+        return predefinedShelfRepository.findByPredefinedShelfNameAndUser(shelfName, userService.getCurrentUser());
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:97]: Could you please add test for this?

File: src/main/java/com/karankumar/bookproject/backend/service/UserService.java [hunk 13]
```diff
@@ -86,6 +87,15 @@ public void register(User user) throws UserAlreadyRegisteredException {
         authenticateUser(user);
     }
 
+    public User getCurrentUser() {
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/UserService.java:90]: Could you please add a test for this?

File: src/main/java/com/karankumar/bookproject/backend/service/UserService.java [hunk 14]
```diff
@@ -86,6 +87,15 @@ public void register(User user) throws UserAlreadyRegisteredException {
...
+        String username = SecurityContextHolder.getContext().getAuthentication().getName();
+        return userRepository.findByUsername(username).orElseThrow();
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/UserService.java:92]: How are we handling the exception? Shall we show a message etc on the UI?

File: src/main/java/com/karankumar/bookproject/backend/service/UserService.java [hunk 15]
```diff
@@ -86,6 +87,15 @@ public void register(User user) throws UserAlreadyRegisteredException {
...
+    }
+
+    public List<User> findAll() {
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/service/UserService.java:95]: Could you please add a test for this change

File: src/main/java/com/karankumar/bookproject/backend/utils/TestData.java [hunk 16]
```diff
@@ -157,13 +157,4 @@ public static PredefinedShelf generateRandomPredefinedShelf(
...
     }
-
-    public static List<PredefinedShelf> createPredefinedShelves(List<Book> books) {
-        return Stream.of(PredefinedShelf.ShelfName.values())
-              .map(shelfName -> {
-                  PredefinedShelf shelf = new PredefinedShelf(shelfName);
-                  shelf.setBooks(new HashSet<>(books));
-                  return shelf;
-              }).collect(Collectors.toList());
-    }
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/utils/TestData.java:168]: Could you please tell why did we remove this?

File: src/test/java/com/karankumar/bookproject/backend/repository/CustomShelfRepositoryTest.java [hunk 17]
```diff
@@ -19,18 +19,20 @@
...
 
 @DataJpaIntegrationTest
 class CustomShelfRepositoryTest {
+    @Autowired private UserRepository userRepository;
     @Autowired private CustomShelfRepository repository;
     private final String customShelf1 = "Test1";
+    private User user;
 
     @Test
     @DisplayName("When shelf exists, findByShelfName returns one shelf")
```
Comment (amit1307) [inline src/test/java/com/karankumar/bookproject/backend/repository/CustomShelfRepositoryTest.java:38]: Could you please add test for checking if there are more than one shelves belonging to different users and we return only the ones belong to the correct users?

File: src/test/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepositoryTest.java [hunk 18]
```diff
@@ -19,39 +19,39 @@
...
+        user = SecurityTestUtils.getTestUser(userRepository);
         repository.saveAll(
             Arrays.stream(PredefinedShelf.ShelfName.values())
-                  .map(PredefinedShelf::new)
+                  .map(shelfName -> new PredefinedShelf(shelfName, user))
                   .collect(Collectors.toList())
         );
     }
 
     @Test
```
Comment (amit1307) [inline src/test/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepositoryTest.java:51]: Could you please add test for checking if there are more than one shelves belonging to different users and we return only the ones belong to the correct users?

File: src/test/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtilsTest.java [hunk 19]
```diff
@@ -86,27 +81,27 @@ public static void setupBeforeAll(@Autowired PredefinedShelfService predefinedSh
...
-        book1 = bookRepository.save(new Book("someTitle", NO_AUTHOR, toReadShelf));
-        book2 = bookRepository.save(new Book("someTitle2", NO_AUTHOR, toReadShelf));
-        book3 = bookRepository.save(new Book("someOtherTitle", NO_AUTHOR, readShelf));
-        book4 = bookRepository.save(new Book("yetAnotherTitle", NO_AUTHOR, didNotFinishShelf));
+    private void createAndSaveBooks() {
+        book1 = bookRepository.save(new Book("someTitle", null, toReadShelf));
+        book2 = bookRepository.save(new Book("someTitle2", null, toReadShelf));
+        book3 = bookRepository.save(new Book("someOtherTitle", null, readShelf));
+        book4 = bookRepository.save(new Book("yetAnotherTitle", null, didNotFinishShelf));
     }
```
Comment (amit1307) [inline src/test/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtilsTest.java:105]: Any reason for changing from more readable NO_AUTHOR to explicitly passing null?

**Review by amit1307 — CHANGES_REQUESTED** [review]

File: src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java (hunk 1 above)
Comment (Burhak) [reply src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java:28]: this method is used instead of `findAll()`

File: src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java (hunk 3 above)
Comment (Burhak) [reply src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java:28]: this method is used instead of `findAll()`

File: src/main/java/com/karankumar/bookproject/backend/utils/TestData.java (hunk 16 above)
Comment (Burhak) [reply src/main/java/com/karankumar/bookproject/backend/utils/TestData.java:168]: It is moved to `PredefinedShelfService` so only this service is responsible for creating new predefined shelves

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java (hunk 8 above)
Comment (Burhak) [reply src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:52]: I don't understand what you propose...

File: src/main/java/com/karankumar/bookproject/backend/service/UserService.java (hunk 14 above)
Comment (Burhak) [reply src/main/java/com/karankumar/bookproject/backend/service/UserService.java:92]: This should **never** happen. It would mean that your app granted access to some user that is actually not in the databse.

**Comment by knjk04** [comment]
@Burhak Thanks for making some changes. I noticed that you marked the conversations as resolved. Please note that in the [resolving conversations](https://github.com/Project-Books/book-project/blob/master/CONTRIBUTING.md#resolving-conversations) section of our contributing guidelines, we explicitly state that this should be left for reviewers. This makes it easier for us to keep track of which of our suggestions have been implemented. 

Feel free, however, to give a thumbs up or a quick message on each to let us know that you've addressed the suggestion

File: src/main/java/com/karankumar/bookproject/backend/service/UserService.java (hunk 14 above)
Comment (amit1307) [reply src/main/java/com/karankumar/bookproject/backend/service/UserService.java:92]: Sure I agree it should never happen but how do we want to react when it happens (for whatever reasons)? Shall we pass the unhandled exception or we should return a proper error message?

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java (hunk 8 above)
Comment (amit1307) [reply src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:52]: I was thinking about if the repository layer would be more appropriate for this change or service layer. But I think it's fine in service layer as well.

File: src/main/java/com/karankumar/bookproject/backend/service/UserService.java (hunk 14 above)
Comment (Burhak) [reply src/main/java/com/karankumar/bookproject/backend/service/UserService.java:92]: If I was a user, and I wanted to perform some action (create a custom shelf for example) and this method had thrown an exception... it would be great to know that my action was not successful. Maybe we could display a message that an unexpected error has occured and log it (later on even send an email to devs or whatever). The same could apply for other unexpected errors (that should never happen), e.g. predefined shelf does not exist.

**Comment by knjk04** [comment]
@Burhak I'll also take a look at this. I want to spend a lot of time on this, so I'll take a look over the weekend

**Comment by knjk04** [comment]
@Burhak Are you also getting a Flyway migration error?

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Invocation of init method failed; nested exception is org.flywaydb.core.api
.FlywayException: Validate failed:
Detected failed migration to version 014 (MAKE SHELVES PER USER)

        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1796) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:595) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:310) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1108) ~[spring-context-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:868) ~[spring-context-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:550) ~[spring-context-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE]
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE]
        at com.karankumar.bookproject.BookProjectApplication.main(BookProjectApplication.java:29) ~[classes!/:0.0.1-SNAPSHOT]
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
        at java.base/java.lang.reflect.Method.invoke(Method.java:567) ~[na:na]
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) ~[book-project-0.0.1-SNAPSHOT.jar:0.0.1-SNAPSHOT]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) ~[book-project-0.0.1-SNAPSHOT.jar:0.0.1-SNAPSHOT]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:51) ~[book-project-0.0.1-SNAPSHOT.jar:0.0.1-SNAPSHOT]
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52) ~[book-project-0.0.1-SNAPSHOT.jar:0.0.1-SNAPSHOT]
Caused by: org.flywaydb.core.api.FlywayException: Validate failed:
Detected failed migration to version 014 (MAKE SHELVES PER USER)

        at org.flywaydb.core.Flyway.doValidate(Flyway.java:292) ~[flyway-core-6.5.1.jar!/:na]
        at org.flywaydb.core.Flyway.access$100(Flyway.java:73) ~[flyway-core-6.5.1.jar!/:na]
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:166) ~[flyway-core-6.5.1.jar!/:na]
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:158) ~[flyway-core-6.5.1.jar!/:na]
        at org.flywaydb.core.Flyway.execute(Flyway.java:527) ~[flyway-core-6.5.1.jar!/:na]
        at org.flywaydb.core.Flyway.migrate(Flyway.java:158) ~[flyway-core-6.5.1.jar!/:na]
        at org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer.afterPropertiesSet(FlywayMigrationInitializer.java:65) ~[spring-boot-autoconfigure-2.2.6.RELEASE.jar!/:2.2.6.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1855) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1792) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE]
        ... 26 common frames omitted

```

File: src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java (hunk 1 above)
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/repository/CustomShelfRepository.java:28]: @amit1307 I think @Burhak is right. This will allow us to get all of the custom shelves that belong to a specific user

File: src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java (hunk 3 above)
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/repository/PredefinedShelfRepository.java:28]: @amit1307 I think this is the same reason for why we need to find all custom shelves for a particular user (we only want to show a user their predefined shelves, not those that belong to other users)

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 20]
```diff
@@ -472,12 +471,6 @@ public void setBook(Book book) {
             return;
         }
 
-        // TODO: this should be removed. A custom shelf should not be mandatory, so it should
-        // be acceptable to the custom shelf to be null
-        if (book.getCustomShelf() == null) {
-            book.setCustomShelf(new CustomShelf("ShelfName"));
-        }
-
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java]: I know this is a code smell, but I'm going to bring this back (using the `CustomShelfService`) as it's a workaround that keeps the app working for now.

File: src/test/java/com/karankumar/bookproject/utils/SecurityTestUtils.java [hunk 21]
```diff
@@ -0,0 +1,24 @@
+package com.karankumar.bookproject.utils;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/utils/SecurityTestUtils.java:1]: Could you please add our [copyright notice](https://github.com/Project-Books/book-project/blob/master/docs/COPYRIGHT) to the top of this file?

File: src/test/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtilsTest.java [hunk 22]
```diff
@@ -49,35 +48,34 @@
 
 @IntegrationTest
 class PredefinedShelfUtilsTest {
-    private static BookRepository bookRepository;
+    @Autowired private BookRepository bookRepository;
+    @Autowired private PredefinedShelfService predefinedShelfService;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtilsTest.java:52]: I agree that we shouldn't have had these in setUp(). I was going to fix this later, so thanks for this. Could you please autowire these in via constructor injection?

File: src/test/java/com/karankumar/bookproject/backend/utils/CustomShelfUtilsTest.java [hunk 23]
```diff
@@ -19,37 +19,36 @@
 
 @IntegrationTest
 class CustomShelfUtilsTest {
-    private static final CustomShelf customShelf1 = new CustomShelf("CustomShelf1");
-    private static final CustomShelf customShelf2 = new CustomShelf("CustomShelf2");
-    private static final CustomShelf customShelfWithNoBooks = new CustomShelf("CustomShelf3");
+    @Autowired private BookService bookService;
+    @Autowired private CustomShelfService customShelfService;
+    @Autowired private PredefinedShelfService predefinedShelfService;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/utils/CustomShelfUtilsTest.java:24]: Could you also please use constructor injection here?

File: src/test/java/com/karankumar/bookproject/backend/statistics/GenreStatisticsTest.java [hunk 24]
```diff
@@ -24,21 +24,22 @@
...
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-
 @IntegrationTest
 class GenreStatisticsTest {
-    private static GenreStatistics genreStatistics;
+    @Autowired private BookService bookService;
+    @Autowired private PredefinedShelfService predefinedShelfService;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/GenreStatisticsTest.java:34]: Could you please also use constructor injection here?

File: src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java [hunk 25]
```diff
@@ -31,36 +30,27 @@
...
-        this.userRepository = userRepository;
-        this.roleRepository = roleRepository;
-    }
-
-    @AfterEach
-    void tearDown() {
-        userRepository.deleteAll();
-        roleRepository.deleteAll();
-    }
-
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java:63]: Could you please undo this change?

File: src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java [hunk 26]
```diff
@@ -31,36 +30,27 @@
 
 import javax.validation.ConstraintViolationException;
 
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static com.karankumar.bookproject.utils.SecurityTestUtils.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java:37]: As per our style guide, we are not using wildcard imports. Could you please use explicit imports, instead?

File: src/test/java/com/karankumar/bookproject/backend/service/PredefinedShelfServiceTest.java [hunk 27]
```diff
@@ -19,22 +19,21 @@
...
 @IntegrationTest
 @DisplayName("PredefinedShelfService should")
 class PredefinedShelfServiceTest {
-    private final PredefinedShelfService predefinedShelfService;
-
-    @Autowired
-    PredefinedShelfServiceTest(PredefinedShelfService predefinedShelfService) {
-        this.predefinedShelfService = predefinedShelfService;
-    }
+    @Autowired private PredefinedShelfService predefinedShelfService;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/PredefinedShelfServiceTest.java:36]: Could you please undo this?

File: src/test/java/com/karankumar/bookproject/backend/service/CustomShelfServiceTest.java [hunk 28]
```diff
@@ -0,0 +1,67 @@
+package com.karankumar.bookproject.backend.service;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/CustomShelfServiceTest.java:1]: Could you please add our [copyright notice](https://github.com/Project-Books/book-project/blob/master/docs/COPYRIGHT) to the top of this file?

File: src/test/java/com/karankumar/bookproject/backend/service/CustomShelfServiceTest.java [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+import static com.karankumar.bookproject.utils.SecurityTestUtils.TEST_USER_NAME;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.SoftAssertions.assertSoftly;
+
+@IntegrationTest
+@Transactional
+@DisplayName("CustomShelfService should")
+class CustomShelfServiceTest {
+    @Autowired private CustomShelfService customShelfService;
+    @Autowired private CustomShelfRepository customShelfRepository;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/CustomShelfServiceTest.java:25]: Could you please use constructor injection here?

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java [hunk 30]
```diff
@@ -54,23 +52,20 @@
     @Autowired private BookService bookService;
     @Autowired private CustomShelfService customShelfService;
     @Autowired private TagService tagService;
+    @Autowired private PredefinedShelfService predefinedShelfService;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:55]: Could you please use constructor injection here? I appreciate there was already field injection, but we have already started refactoring our tests to move to constructor injection

File: src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java [hunk 31]
```diff
@@ -32,25 +32,14 @@
...
+    @Autowired private TagService tagService;
+    @Autowired private PredefinedShelfService predefinedShelfService;
 
-    private static Book testBook;
-    private static Tag testTag;
-
-    @Autowired
-    public BookTest(PredefinedShelfService predefinedShelfService,
-                    BookService bookService,
-                    TagService tagService) {
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java:45]: Could you please undo this? We still would like to use constructor injection

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 32]
```diff
@@ -21,40 +21,30 @@
...
-    private final AuthorService authorService;
-    private static PredefinedShelf toRead;
+    private Book testBook1;
+    private Book testBook2;
+    private PredefinedShelf toRead;
 
-    @Autowired
-    public AuthorTest(PredefinedShelfService predefinedShelfService,
-                      BookService bookService,
-                      AuthorService authorService) {
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:47]: Could you please bring back constructor injection?

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java (hunk 8 above)
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:52]: @Burhak For consistency, could you please pass in the `UserRepository` here?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks a lot for working on this! 

I've left some feedback. It's mostly stylistic suggestions.

**Comment by knjk04** [comment]
Also, could you please resolve the merge conflict and rebase from master?

**Comment by Burhak** [comment]
@knjk04 We cannot pass `UserRepository ` to `PredefinedShelfService` instead of `UserService` because we are getting logged in user from user service.

**Comment by knjk04** [comment]
@Burhak Thanks for the clarification. Ignore me then :)

File: src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java (hunk 25 above)
Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java]: Thanks for bringing back constructor injection. When you're ready for a review, please also ensure `tearDown()` is brought back

Comment (Burhak) [reply src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java]: I annotated this test with `@Transactional` so `tearDown()` is not needed anymore because the transaction is rolled back after each test.

**Comment by Burhak** [comment]
@knjk04 Conflicts resolved and I'm ready for another review (if `UserServiceTest` is okay).

**Comment by knjk04** [comment]
@Burhak Thanks a lot for working on this! This PR is great!

**Comment by knjk04** [comment]
@Burhak If you're interested, feel free to join our [Slack](https://join.slack.com/t/teambookproject/shared_invite/zt-i7lept44-rgJ9yB0A2vedJTLyyfkjKQ) workspace. We could definitely do with someone with your skill helping to shape the project
